src/solvers/dg_p4est/dg_3d.jl

```diff
@@ -0,0 +1,407 @@
+                                               element]
+      end
+    end
+  end
+
+  return nothing
+end
+
+
+function calc_boundary_flux!(cache, t, boundary_condition, boundary_indexing,
+                             mesh::P4estMesh{3},
+                             equations, surface_integral, dg::DG)
+  @unpack boundaries = cache
+  @unpack surface_flux_values, node_coordinates = cache.elements
+  @unpack surface_flux = surface_integral
+
+  size_ = (nnodes(dg), nnodes(dg), nnodes(dg))
+
+  @threaded for local_index in eachindex(boundary_indexing)
+    # Use the local index to get the global boundary index from the pre-sorted list
+    boundary = boundary_indexing[local_index]
+
+    element       = boundaries.element_ids[boundary]
+    node_indices  = boundaries.node_indices[boundary]
+    direction     = indices2direction(node_indices)
+
+    # Use Tuple `node_indices` and `evaluate_index` to access node indices
+    # at the correct face and in the correct orientation to get normal vectors
+    for j in eachnode(dg), i in eachnode(dg)
+      node_i = evaluate_index(node_indices, size_, 1, i, j)
+      node_j = evaluate_index(node_indices, size_, 2, i, j)
+      node_k = evaluate_index(node_indices, size_, 3, i, j)
+
+      # Extract solution data from boundary container
+      u_inner = get_node_vars(boundaries.u, equations, dg, i, j, boundary)
+
+      # Outward-pointing normal vector
+      normal_vector = get_normal_vector(direction, cache, node_i, node_j, node_k, element)
+
+      # Coordinates at boundary node
+      x = get_node_coords(node_coordinates, equations, dg, node_i, node_j, node_k, element)
+
+      flux_ = boundary_condition(u_inner, normal_vector, x, t, surface_flux, equations)
+
+      # Use Tuple `node_indices` and `evaluate_index_surface` to copy flux
+      # to left and right element storage in the correct orientation
+      for v in eachvariable(equations)
+        surf_i = evaluate_index_surface(node_indices, size_, 1, i, j)
+        surf_j = evaluate_index_surface(node_indices, size_, 2, i, j)
+        surface_flux_values[v, surf_i, surf_j, direction, element] = flux_[v]
+      end
+    end
+  end
+end
+
+
+function prolong2mortars!(cache, u,
+                          mesh::P4estMesh{3}, equations,
+                          mortar_l2::LobattoLegendreMortarL2,
+                          surface_integral, dg::DGSEM)
+  # temporary buffer for projections
+  @unpack fstar_tmp_threaded = cache
+  @unpack element_ids, node_indices = cache.mortars
+
+  size_ = (nnodes(dg), nnodes(dg), nnodes(dg))
+
+  @threaded for mortar in eachmortar(dg, cache)
+    fstar_tmp = fstar_tmp_threaded[Threads.threadid()]
+
+    small_indices = node_indices[1, mortar]
+    large_indices = node_indices[2, mortar]
+
+    # Copy solution small to small
+    for pos in 1:4, j in eachnode(dg), i in eachnode(dg)
+      for v in eachvariable(equations)
```

Comment:
Sorry for not being clearer before: IIRC, so far we have always used a separate for loop for each variable, *except* for the loops over nodes. Thus, this loop (and similar ones) should be written as
```suggestion
    for pos in 1:4
      for j in eachnode(dg), i in eachnode(dg)
        for v in eachvariable(equations)
```
with one more indentation level inside the inner-most loop.

Reply:
Please correct me if I'm wrong, though!

Reply:
So many loops

Reply:
Yes, but IMHO much easier to read & comprehend if each is on a separate line ;-)